diff --git a/browser/components/preferences/jar.mn b/browser/components/preferences/jar.mn
index 6fd90b4..43d2421 100644
--- a/browser/components/preferences/jar.mn
+++ b/browser/components/preferences/jar.mn
@@ -12,6 +12,7 @@ browser.jar:
    content/browser/preferences/home.js
    content/browser/preferences/search.js
    content/browser/preferences/privacy.js
+   content/browser/preferences/sentinel.js
    content/browser/preferences/sync.js
    content/browser/preferences/moreFromMozilla.js
    content/browser/preferences/fxaPairDevice.xhtml
diff --git a/browser/components/preferences/preferences.js b/browser/components/preferences/preferences.js
index c76a064..845d0dc 100644
--- a/browser/components/preferences/preferences.js
+++ b/browser/components/preferences/preferences.js
@@ -7,6 +7,7 @@
 /* import-globals-from home.js */
 /* import-globals-from search.js */
 /* import-globals-from privacy.js */
+/* import-globals-from sentinel.js */
 /* import-globals-from sync.js */
 /* import-globals-from moreFromMozilla.js */
 /* import-globals-from findInPage.js */
diff --git a/browser/components/preferences/preferences.xhtml b/browser/components/preferences/preferences.xhtml
index 1b8c381..b5531ac 100644
--- a/browser/components/preferences/preferences.xhtml
+++ b/browser/components/preferences/preferences.xhtml
@@ -35,6 +35,7 @@
   <link rel="stylesheet" href="chrome://browser/skin/preferences/search.css" />
   <link rel="stylesheet" href="chrome://browser/content/usercontext/usercontext.css" />
   <link rel="stylesheet" href="chrome://browser/skin/preferences/privacy.css" />
+  <link rel="stylesheet" href="chrome://browser/skin/preferences/sentinel.css" />
   <link rel="stylesheet" href="chrome://browser/content/preferences/widgets/setting-group.css" />
   <link rel="stylesheet" href="chrome://browser/content/preferences/widgets/setting-pane.css" />
   <link rel="stylesheet" href="chrome://browser/content/preferences/widgets/setting-control.css" />
@@ -170,10 +171,10 @@
         iconsrc="chrome://browser/skin/translations.svg"
         data-l10n-id="pane-languages-title2">
       </html:moz-page-nav-button>
-      <html:moz-page-nav-button id="category-ai-features"
-        view="paneAi"
-        iconsrc="chrome://global/skin/icons/highlights.svg"
-        data-l10n-id="pane-ai-controls-title2">
+      <html:moz-page-nav-button id="category-sentinel"
+        view="paneSentinel"
+        iconsrc="chrome://browser/skin/preferences/category-sentinel.svg"
+        data-l10n-id="pane-sentinel-title2">
       </html:moz-page-nav-button>
       <html:moz-page-nav-button id="category-permissions-data"
         view="panePermissionsData"
@@ -246,6 +247,7 @@
 #include home.inc.xhtml
 #include search.inc.xhtml
 #include privacy.inc.xhtml
+#include sentinel.inc.xhtml
 #include sync.inc.xhtml
 #include moreFromMozilla.inc.xhtml
         </vbox>
diff --git a/browser/themes/shared/jar.inc.mn b/browser/themes/shared/jar.inc.mn
index c2bebf7e96..62ab632cec 100644
--- a/browser/themes/shared/jar.inc.mn
+++ b/browser/themes/shared/jar.inc.mn
@@ -114,6 +114,7 @@
   skin/classic/browser/preferences/android-menu.svg            (../shared/preferences/android-menu.svg)
   skin/classic/browser/preferences/category-general.svg        (../shared/preferences/category-general.svg)
   skin/classic/browser/preferences/category-privacy-security.svg    (../shared/preferences/category-privacy-security.svg)
+  skin/classic/browser/preferences/category-sentinel.svg    (../shared/preferences/category-sentinel.svg)
   skin/classic/browser/preferences/category-search.svg         (../shared/preferences/category-search.svg)
   skin/classic/browser/preferences/category-sync.svg           (../shared/preferences/category-sync.svg)
   skin/classic/browser/preferences/containers.css              (../shared/preferences/containers.css)
@@ -131,6 +132,7 @@
   skin/classic/browser/preferences/vpn-logo.svg                (../shared/preferences/vpn-logo.svg)
   skin/classic/browser/preferences/search.css                  (../shared/preferences/search.css)
   skin/classic/browser/preferences/siteDataSettings.css        (../shared/preferences/siteDataSettings.css)
+  skin/classic/browser/preferences/sentinel.css               (../shared/preferences/sentinel.css)
   skin/classic/browser/translations/panel.css                  (../shared/translations/panel.css)
   skin/classic/browser/fxa/fxa-spinner.svg                     (../shared/fxa/fxa-spinner.svg)
   skin/classic/browser/fxa/sync-illustration.svg               (../shared/fxa/sync-illustration.svg)
diff --git a/browser/components/preferences/main.js b/browser/components/preferences/main.js
index a8de115..f6885f1 100644
--- a/browser/components/preferences/main.js
+++ b/browser/components/preferences/main.js
@@ -571,6 +571,79 @@ SettingGroupManager.registerGroups({
   startup: createStartupConfig(
     Services.prefs.getBoolPref("browser.settings-redesign.enabled", false)
   ),
+  sentinelBehavior: {
+    l10nId: "sentinel-general-heading2",
+    headingLevel: 2,
+    items: [
+      {
+        id: "sentinelExtensionUpdate",
+        l10nId: "sentinel-extension-update-checkbox2",
+      },
+      {
+        id: "sentinelSync",
+        l10nId: "sentinel-sync-checkbox2",
+      },
+      {
+        id: "sentinelMiddleClick",
+        l10nId: "sentinel-autocopy-checkbox2",
+      },
+      {
+        id: "sentinelNova",
+        l10nId: "sentinel-nova-checkbox2",
+      },
+      {
+        id: "sentinelUserChrome",
+        l10nId: "sentinel-styling-checkbox2",
+      },
+    ],
+  },
+  sentinelNetworking: {
+    l10nId: "sentinel-network-heading2",
+    headingLevel: 2,
+    items: [
+      {
+        id: "sentinelIPv6",
+        l10nId: "sentinel-ipv6-checkbox2",
+      }
+    ],
+  },
+  sentinelPrivacy: {
+    l10nId: "sentinel-privacy-heading2",
+    headingLevel: 2,
+    items: [
+      {
+        id: "sentinelJIT",
+        l10nId: "sentinel-jit-checkbox2",
+        supportPage: "jit-toggle",
+      },
+      {
+        id: "sentinelCrossOrigin",
+        l10nId: "sentinel-xorigin-ref-checkbox2",
+      },
+    ],
+  },
+  sentinelFingerprinting: {
+    l10nId: "sentinel-broken-heading2",
+    headingLevel: 2,
+    items: [
+      {
+        id: "sentinelRFP",
+        l10nId: "sentinel-rfp-checkbox2",
+      },
+      {
+        id: "sentinelLetterboxing",
+        l10nId: "sentinel-letterboxing-checkbox2",
+      },
+      {
+        id: "sentinelWebGLPrompt",
+        l10nId: "sentinel-webgl-checkbox2",
+      },
+      {
+        id: "sentinelWebGLPromptHide",
+        l10nId: "sentinel-webgl-prompt-checkbox2",
+      },
+    ],
+  },
 });
 
 /**
